torchao/prototype/dtypes/uint2.py

```diff
@@ -0,0 +1,233 @@
+def view_dtype(func, args, kwargs):
+    tensor, dtype = args
+    if dtype is torch.uint8:
+        return unpack_uint2(tensor.elem).to(torch.uint8)
+    raise NotImplementedError(f"view {dtype} not supported")
+
+@implements([torch.ops.aten.clone.default])
+def clone(func, args, kwargs):
+    tensor = args[0]
+    return UInt2Tensor(tensor.elem.clone())
+
+@implements([torch.ops.aten._unsafe_view.default])
+def unsafe_view(func, args, kwargs):
+    tensor, size = args
+    size = utils.infer_size(size, tensor.numel())
+    assert not kwargs
+    dsize = down_size(size)
+    reshaped_elem = tensor.elem.view(dsize)
+    return UInt2Tensor(reshaped_elem)
+
+@implements([torch.ops.aten.unbind.int])
+def unbind(func, args, kwargs):
+    tensor, dim = fill_defaults(args, 2, [0])
+    if dim != tensor.dim() - 1:
+        raise NotImplementedError(f"unbind dim={dim}")
+    else:
+        x = tensor.elem.to(torch.uint8).unbind(dim)
+        return x
+
+@implements([torch.ops.aten._to_copy.default])
+def to_copy(func, args, kwargs):
+    (tensor,) = args
+    dtype = kwargs["dtype"]
+    if dtype == torch.uint8:
+        return unpack_uint2(tensor.elem).view(tensor.shape).view(torch.uint8)
+    elif dtype in (torch.float, torch.float16, torch.bfloat16, torch.int16, torch.int32, torch.int64):
+        return tensor.to(torch.uint8).to(dtype)
+    elif isinstance(tensor, UInt2Tensor):
+        return tensor
+    raise NotImplementedError(f"to_copy {dtype} not supported")
+
+@implements([torch.ops.aten.select.int])
+def select(func, args, kwargs):
+    tensor, dim, index = args
+    if dim != tensor.dim() - 1:
+        selected_elem = tensor.elem.select(dim, index)
+        return UInt2Tensor(selected_elem)
+    else:
+        raise NotImplementedError(f"select dim={dim}")
+
+@implements([torch.ops.aten.reshape.default])
+def reshape(func, args, kwargs):
+    tensor, size = args
+    size = utils.infer_size(size, tensor.numel())
+    assert not kwargs
+    dsize = down_size(size)
+    reshaped_elem = tensor.elem.view(dsize)
+    return UInt2Tensor(reshaped_elem)
+
+def slice_tensor(func, args, kwargs):
+    tensor, dim, start, end, step = fill_defaults(args, 5, [0, None, None, 1])
+    if dim == tensor.dim() - 1:
+        if step != 1:
+            raise NotImplementedError(f"slice step={step}")
+        assert start % 4 == 0, start
+        assert end is None or end % 4 == 0, end
+        end = end if end is not None else tensor.shape[dim]
+        sliced_elem = tensor.elem[..., start // 4 : end // 4 : step]
+        return UInt2Tensor(sliced_elem)
+    else:
+        sliced_elem = tensor.elem[..., start:end:step]
+        return UInt2Tensor(sliced_elem)
+
+@implements([torch.ops.aten.equal.default])
+def equal(func, args, kwargs):
+    tensor, other = args
+    return torch.equal(tensor.elem, other.elem)
+
+@implements([torch.ops.aten.detach.default])
+def detach(func, args, kwargs):
+    (tensor,) = args
+    return tensor.elem.detach()
```

Comment:
sidecar comment:

- most of your view op implementations above wrap the output back into a `Uint2Tensor` ("propagating" the subclass-ness through the model when views are encountered)

- `detach` is just another view op

- so you probably want your impl for detach to wrap the output in your subclass?

Reply:
Yes that is correct. Fixed it